accounts/abi/bind/backends/blockchain.go

```diff
@@ -0,0 +1,142 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The klaytn library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the klaytn library. If not, see <http://www.gnu.org/licenses/>.
+
+package backends
+
+import (
+	"context"
+	"math/big"
+
+	"github.com/klaytn/klaytn"
+	"github.com/klaytn/klaytn/accounts/abi/bind"
+	"github.com/klaytn/klaytn/blockchain"
+	"github.com/klaytn/klaytn/blockchain/state"
+	"github.com/klaytn/klaytn/blockchain/types"
+	"github.com/klaytn/klaytn/blockchain/vm"
+	"github.com/klaytn/klaytn/common"
+	"github.com/klaytn/klaytn/params"
+)
+
+// Maintain separate minimal interfaces of blockchain.BlockChain because ContractBackend are used
+// in various situations. BlockChain instances are often passed down as different interfaces such as
+// consensus.ChainReader, governance.blockChain, work.BlockChain.
+type blockChainForCaller interface {
+	// Required by NewEVMContext
+	blockchain.ChainContext
+
+	// Below is a subset of consensus.ChainReader
+	// Only using the vocabulary of consensus.ChainReader for potential
+	// usability within consensus package.
+	Config() *params.ChainConfig
+	CurrentHeader() *types.Header
+	GetHeaderByNumber(number uint64) *types.Header
+	GetBlock(hash common.Hash, number uint64) *types.Block
+	State() (*state.StateDB, error)
+	StateAt(root common.Hash) (*state.StateDB, error)
+}
+
+// BlockchainContractCaller implements bind.ContractCaller, based on
+// a user-supplied blockchain.BlockChain instance.
+// Its intended purpose is reading system contracts during block processing.
+//
+// Note that SimulatedBackend creates a new temporary BlockChain for testing,
+// whereas BlockchainContractCaller uses an existing BlockChain with existing database.
+type BlockchainContractCaller struct {
+	bc blockChainForCaller
+}
+
+// This nil assignment ensures at compile time that BlockchainContractCaller implements bind.ContractCaller.
+var _ bind.ContractCaller = (*BlockchainContractCaller)(nil)
+
+func NewBlockchainContractCaller(bc blockChainForCaller) *BlockchainContractCaller {
+	return &BlockchainContractCaller{
+		bc: bc,
+	}
+}
+
+func (b *BlockchainContractCaller) CodeAt(ctx context.Context, account common.Address, blockNumber *big.Int) ([]byte, error) {
+	if _, state, err := b.getBlockAndState(blockNumber); err != nil {
+		return nil, err
+	} else {
+		return state.GetCode(account), nil
+	}
+}
+
+// Executes a read-only function call with respect to the latest state.
+//
+// Returns call result in []byte.
+// Returns error when:
+// - a past (i.e. not latest) blockNumber is given
```

Comment:
why is this? does this mean GetStakingInfo or ParamsAt with past block cannot be handled?

Reply:
It is a outdated comment, which should be updated.

The original implementation of BlockChainContractCaller did not have StateAt(). Now it has StateAt(), and past block can be handled. I've updated the comment accordingly.

Thanks!